torch_np/_unary_ufuncs.py

```diff
@@ -1,24 +1,7 @@
-import functools
-import torch
+from ._decorators import deco_unary_ufunc_from_impl
+from ._detail import _ufunc_impl
```

Comment:
I don't quite understand this file. Should you do `from ._detail._ufunc_impl import *` or so? How come you are exporing all those functions in `__all__` but they are not defined at all in this file?

Reply:
doing `from ._detail._ufunc_impl import *` would be messy because the names are the same. The logic is roughly:
- `torch_np/__init__.py` does `from _unary_ufuncs import *`, which brings public names, e.g. `sin`
- `_unary_ufuncs.sin` receives wrapper arrays and other args (dtypes, casting etc). It extract the tensor from the wrapper array, and delegates to `_ufunc_impl.sin`
- `_ufunc_impl.sin` receives a tensor and other args, handles dtypes, casts and broadcasting (in the future, also `where=...)`, and delegates to `torch.sin`

`__all__` here is a technical implement for smoke testing in https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/refactor/torch_np/tests/test_basic.py#L29

Can rename `__all__` to something else if that's helpful?

Reply:
Note that `__all__` has a very specific meaning in Python (see e.g. https://stackoverflow.com/a/64130/5280578). It should be clear after reading that why using `__all__` here is incorrect.

What I would suggest doing is:
- Set the `__all__` variable correctly everywhere to express what is it that you mean to export publicly from that file.
- After doing that, pulling the relevant functions from a given file (in that `test_basic.py` or elsewhere) should be trivial.

Reply:
Well, one other meaning of  `__all__` is "these names are meant to be public even though star imports are bad idea and we do not recommend them". Still other use is to do a star import from a private submodule in `__init__.py`  --- which is exactly the thing here. 

I am going to push  setting `__all__` correctly to a next refactoring pass.

Reply:
>  Still other use is to do a star import from a private submodule in `__init__.py` --- which is exactly the thing here.

This relates to why this felt weird to me in the first place. You are setting the `__all__` attribute without making those functions available from that file, which is odd. We can discuss it in tomorrow's meeting, and I may very well be in the wrong here, but I've certainly never seen this pattern before.